### PR TITLE
Add `CQM.substitute_self_loops` method

### PR DESCRIPTION
**Pull request by arcondello**
Also some other misc bug fixes.

### CODE REVIEW

File: dimod/constrained.py [hunk 1]
```diff
@@ -599,6 +600,80 @@ def set_objective(self, objective: Union[BinaryQuadraticModel, QuadraticModel]):
...
+
+            qm.add_quadratic(u, new, bias)
+            qm.remove_interaction(u, u)
+
+    def substitute_self_loops(self) -> Dict[Variable, Variable]:
+        """Replace any integer self-loops in the objective or constraints.
+
+        A self-loop :math:`i^2` is removed by introducing a new
+        variable :math:`j`, adding a constraint :math:`j == i`, and then
+        substitute :math:`i^2 == i*j`.
```
Comment (JoelPasvolsky) [inline dimod/constrained.py:654]: Maybe more accurate (and without the switch in tense in the last clause): "Self-loop :math:`i^2` is removed by introducing a new variable :math:`j` with interaction :math:`i*j` and adding constraint :math:`j == i`."

File: dimod/constrained.py [hunk 2]
```diff
@@ -599,6 +600,80 @@ def set_objective(self, objective: Union[BinaryQuadraticModel, QuadraticModel]):
...
+
+        Acts on the objective and constraints in-place.
+
+        Returns:
+            A mapping from the integer variable labels to their introduced
+            counterparts. The same label is used for the constraint that
+            enforces the relationship.
```
Comment (JoelPasvolsky) [inline dimod/constrained.py:661]: "The same label is used for the constraint that enforces the relationship." Unclear.

File: releasenotes/notes/add-QM.remove_interaction-e97a7ac91db2e2b8.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+features:
+  - |
+    Implement ``QuadraticModel.remove_interaction`` method, previously it
+    raised a ``NotImplementedError`` error.
```
Comment (JoelPasvolsky) [inline releasenotes/notes/add-QM.remove_interaction-e97a7ac91db2e2b8.yaml:5]: If you're splicing with a comma, then "method, which previously raised"

Comment (arcondello) [reply releasenotes/notes/add-QM.remove_interaction-e97a7ac91db2e2b8.yaml:5]: Too late. It's in the git history like that forever. Haunting you.

Comment (JoelPasvolsky) [reply releasenotes/notes/add-QM.remove_interaction-e97a7ac91db2e2b8.yaml:5]: Well, you're stuck forever with https://github.com/dwavesystems/dimod/pull/947#issuecomment-896239578

**Pull request merged**